File: rust/src/leetcode/Q124_binary_tree_maximum_path_sum.rs
/*
 * @lc app=leetcode id=124 lang=rust
 *
 * [124] Binary Tree Maximum Path Sum
 */

// Definition for a binary tree node.
#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<Rc<RefCell<TreeNode>>>,
    pub right: Option<Rc<RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}

// @lc code=start
use std::cell::RefCell;
use std::cmp::max;
use std::rc::Rc;

impl Solution {
    fn contained_and_included_path(root: Option<Rc<RefCell<TreeNode>>>) -> (i32, i32) {
        if let Some(node) = root {
            let (left_contained, left_included) =
                Self::contained_and_included_path(node.borrow().left.clone());
            let (right_contained, right_included) =
                Self::contained_and_included_path(node.borrow().right.clone());

            let left_updated: i32 = max(left_included, 0);
            let right_updated: i32 = max(right_included, 0);

            let sub_contained: i32 = max(left_contained, right_contained);
            let contaiend: i32 = max(
                sub_contained,
                left_updated + right_updated + node.borrow().val,
            );

            let included: i32 = max(left_updated, right_updated) + node.borrow().val;

            (contaiend, included)
        } else {
            (i32::MIN, 0)
        }
    }

    pub fn max_path_sum(root: Option<Rc<RefCell<TreeNode>>>) -> i32 {
        Self::contained_and_included_path(root).0
    }
}
// @lc code=end

File: python/Q124_binary_tree_maximum_path_sum.py
#
# @lc app=leetcode id=124 lang=python3
#
# [124] Binary Tree Maximum Path Sum
#


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None) -> None:
        self.val: int = val
        self.left: TreeNode | None = left
        self.right: TreeNode | None = right


# @lc code=start
class Solution:
    def max_path(self, root: TreeNode | None) -> tuple[int, int]:
        if not root:
            return 0, -3 * 10**7

        left_connected, left_max = self.max_path(root.left)
        right_connected, right_max = self.max_path(root.right)

        left_updated: int = left_connected if left_connected > 0 else 0
        right_updated: int = right_connected if right_connected > 0 else 0

        connected_max: int = (
            left_updated if left_updated > right_updated else right_updated
        )

        connected_max += root.val

        previous_path: int = left_max if left_max > right_max else right_max
        current_path: int = left_updated + root.val + right_updated
        max_path: int = previous_path if previous_path > current_path else current_path

        return connected_max, max_path

    def maxPathSum(self, root: TreeNode | None) -> int:
        return self.max_path(root)[1]


# @lc code=end
